modules/programs/firefox/profiles/search.nix

```diff
@@ -0,0 +1,251 @@
+          config = configs.${name} or {
+            _name = name;
+            _isAppProvided = true;
+            _metaData = { };
+          };
+        in config // { _metaData = config._metaData // { inherit order; }; };
+
+      engineConfigsWithoutOrder = attrValues (removeAttrs configs config.order);
+
+      sortedEngineConfigs = (imap buildEngineConfigWithOrder config.order)
+        ++ engineConfigsWithoutOrder;
+    in sortedEngineConfigs;
+
+  engineInput = config.engines // {
+    # Infer config.default as an app provided
+    # engine if it's not in config.engines
+    ${config.default} = config.engines.${config.default} or { };
+  } // {
+    ${config.privateDefault} = config.engines.${config.privateDefault} or { };
+  };
+
+  settings = {
+    version = 6;
+    engines = sortEngineConfigs (mapAttrs buildEngineConfig engineInput);
+
+    metaData = optionalAttrs (config.default != null) {
+      current = config.default;
+      hash = "@hash@";
+    } // optionalAttrs (config.privateDefault != null) {
+      private = config.privateDefault;
+      privateHash = "@privateHash@";
+    } // {
+      useSavedOrder = config.order != [ ];
+    };
+  };
+
+  # Home Manager doesn't circumvent user consent and isn't acting
+  # maliciously. We're modifying the search outside of the browser, but
+  # a claim by Mozilla to remove this would be very anti-user, and
+  # is unlikely to be an issue for our use case.
+  disclaimer = "By modifying this file, I agree that I am doing so "
+    + "only within ${appName} itself, using official, user-driven search "
+    + "engine selection processes, and in a way which does not circumvent "
+    + "user consent. I acknowledge that any attempt to change this file "
+    + "from outside of ${appName} is a malicious act, and will be responded "
+    + "to accordingly.";
+
+  salt = if config.default != null then
+    profilePath + config.default + disclaimer
+  else
+    null;
+
+  privateSalt = if config.privateDefault != null then
+    profilePath + config.privateDefault + disclaimer
+  else
+    null;
+
+  file = pkgs.runCommand "search.json.mozlz4" {
+    nativeBuildInputs = with pkgs; [ mozlz4a openssl ];
+    json = builtins.toJSON settings;
+    inherit salt privateSalt;
+  } ''
+    if [[ -n $salt ]]; then
+      export hash=$(echo -n "$salt" | openssl dgst -sha256 -binary | base64)
+      export privateHash=$(echo -n "$privateSalt" | openssl dgst -sha256 -binary | base64)
+      mozlz4a <(substituteStream json search.json.in --subst-var hash --subst-var privateHash) "$out"
+    else
+      mozlz4a <(echo "$json") "$out"
+    fi
+  '';
+in {
+  imports = [ (pkgs.path + "/nixos/modules/misc/meta.nix") ];
+
+  meta.maintainers = with maintainers; [ kira-bruneau ];
+
+  options = {
+    enable = mkOption {
+      type = with types; bool;
+      default = config.default != null || config.privateDefault != null
+        || config.order != [ ] || config.engines != { };
+      internal = true;
+    };
```

Comment:
I added this internal `enable` option to make the implicit enable behaviour reusable, but I think ideally users should be required to set `enable` explicitly. I just didn't want to introduce any breaking changes in this PR.

I plan on folding `force` into `enable` in a follow-up PR.